File: packages/ensemble/src/stacking.js
import {
  encodeBundle, decodeBundle, register, load as registryLoad,
  normalizeX, normalizeY, accuracy, r2Score,
  stratifiedKFold, kFold,
  ValidationError, NotFittedError, DisposedError
} from '@wlearn/core'

const TYPE_ID_CLS = 'wlearn.ensemble.stacking.classifier@1'
const TYPE_ID_REG = 'wlearn.ensemble.stacking.regressor@1'
let _registered = false

export class StackingEnsemble {
  #baseSpecs      // [name, Class, params][]
  #metaSpec       // [name, Class, params]
  #baseModels     // fitted base model instances (on full data)
  #metaModel      // fitted meta-model instance
  #cv
  #task
  #passthrough
  #seed
  #classes
  #nClasses
  #nMetaCols
  #fitted = false
  #disposed = false

  constructor(params) {
    this.#baseSpecs = params.estimators || []
    this.#metaSpec = params.finalEstimator || null
    this.#cv = params.cv || 5
    this.#task = params.task || 'classification'
    this.#passthrough = params.passthrough || false
    this.#seed = params.seed ?? 42
    this.#baseModels = null
    this.#metaModel = null
    this.#classes = null
    this.#nClasses = 0
    this.#nMetaCols = 0
    StackingEnsemble._register()
  }

  static async create(params = {}) {
    return new StackingEnsemble(params)
  }

  #ensureAlive() {
    if (this.#disposed) throw new DisposedError('StackingEnsemble has been disposed.')
  }

  #ensureFitted() {
    this.#ensureAlive()
    if (!this.#fitted) throw new NotFittedError('StackingEnsemble is not fitted. Call fit() first.')
  }

  async fit(X, y) {
    this.#ensureAlive()
    if (!this.#metaSpec) {
      throw new ValidationError('StackingEnsemble requires a finalEstimator')
    }

    const Xn = normalizeX(X)
    const yn = normalizeY(y)
    const n = Xn.rows

    // Discover classes
    if (this.#task === 'classification') {
      const labelSet = new Set()
      for (let i = 0; i < yn.length; i++) labelSet.add(yn[i])
      this.#classes = new Int32Array([...labelSet].sort((a, b) => a - b))
      this.#nClasses = this.#classes.length
    }

    // Generate folds
    const folds = this.#task === 'classification'
      ? stratifiedKFold(yn, this.#cv, { shuffle: true, seed: this.#seed })
      : kFold(n, this.#cv, { shuffle: true, seed: this.#seed })

    // Step 1: Generate OOF predictions for each base model
    const nBase = this.#baseSpecs.length
    const colsPerModel = this.#task === 'classification' ? this.#nClasses : 1
    const oofCols = nBase * colsPerModel
    const oofData = new Float64Array(n * oofCols)

    for (let b = 0; b < nBase; b++) {
      const [, EstClass, params] = this.#baseSpecs[b]
      for (const { train, test } of folds) {
        const Xtrain = _subsetX(Xn, train)
        const ytrain = _subsetY(yn, train)
        const Xtest = _subsetX(Xn, test)

        const model = await EstClass.create(params || {})
        try {
          model.fit(Xtrain, ytrain)
          if (this.#task === 'classification') {
            const proba = model.predictProba(Xtest)
            for (let i = 0; i < test.length; i++) {
              const row = test[i]
              for (let c = 0; c < this.#nClasses; c++) {
                oofData[row * oofCols + b * colsPerModel + c] = proba[i * this.#nClasses + c]
              }
            }
          } else {
            const preds = model.predict(Xtest)
            for (let i = 0; i < test.length; i++) {
              oofData[test[i] * oofCols + b] = preds[i]
            }
          }
        } finally {
          model.dispose()
        }
      }
    }

    // Step 2: Build meta-feature matrix
    let metaX
    if (this.#passthrough) {
      this.#nMetaCols = oofCols + Xn.cols
      const metaData = new Float64Array(n * this.#nMetaCols)
      for (let i = 0; i < n; i++) {
        // OOF predictions
        metaData.set(
          oofData.subarray(i * oofCols, (i + 1) * oofCols),
          i * this.#nMetaCols
        )
        // Original features
        metaData.set(
          Xn.data.subarray(i * Xn.cols, (i + 1) * Xn.cols),
          i * this.#nMetaCols + oofCols
        )
      }
      metaX = { data: metaData, rows: n, cols: this.#nMetaCols }
    } else {
      this.#nMetaCols = oofCols
      metaX = { data: oofData, rows: n, cols: oofCols }
    }

    // Step 3: Train base models on full data
    this.#baseModels = []
    for (const [, EstClass, params] of this.#baseSpecs) {
      const model = await EstClass.create(params || {})
      model.fit(Xn, yn)
      this.#baseModels.push(model)
    }

    // Step 4: Train meta-model on OOF features
    const [, MetaClass, metaParams] = this.#metaSpec
    this.#metaModel = await MetaClass.create(metaParams || {})
    this.#metaModel.fit(metaX, yn)

    this.#fitted = true
    return this
  }

  predict(X) {
    this.#ensureFitted()
    const metaX = this.#buildMetaFeatures(X)
    return this.#metaModel.predict(metaX)
  }

  predictProba(X) {
    this.#ensureFitted()
    if (this.#task !== 'classification') {
      throw new ValidationError('predictProba is only available for classification')
    }
    if (typeof this.#metaModel.predictProba !== 'function') {
      throw new ValidationError('Meta-model does not support predictProba')
    }
    const metaX = this.#buildMetaFeatures(X)
    return this.#metaModel.predictProba(metaX)
  }

  score(X, y) {
    this.#ensureFitted()
    const preds = this.predict(X)
    const yn = normalizeY(y)
    return this.#task === 'classification' ? accuracy(yn, preds) : r2Score(yn, preds)
  }

  save() {
    this.#ensureFitted()
    const typeId = this.#task === 'classification' ? TYPE_ID_CLS : TYPE_ID_REG
    const manifest = {
      typeId,
      params: {
        task: this.#task,
        cv: this.#cv,
        passthrough: this.#passthrough,
        seed: this.#seed,
        estimatorNames: this.#baseSpecs.map(s => s[0]),
        metaName: this.#metaSpec[0],
        classes: this.#classes ? [...this.#classes] : null,
        nMetaCols: this.#nMetaCols,
      },
    }
    const artifacts = this.#baseModels.map((model, i) => ({
      id: this.#baseSpecs[i][0],
      data: model.save(),
      mediaType: 'application/x-wlearn-bundle',
    }))
    artifacts.push({
      id: this.#metaSpec[0],
      data: this.#metaModel.save(),
      mediaType: 'application/x-wlearn-bundle',
    })
    return encodeBundle(manifest, artifacts)
  }

  static async load(bytes) {
    const { manifest, toc, blobs } = decodeBundle(bytes)
    return StackingEnsemble._loadFromParts(manifest, toc, blobs)
  }

  dispose() {
    if (this.#disposed) return
    this.#disposed = true
    if (this.#baseModels) {
      for (const m of this.#baseModels) m.dispose()
    }
    if (this.#metaModel) this.#metaModel.dispose()
  }

  getParams() {
    return {
      task: this.#task,
      cv: this.#cv,
      passthrough: this.#passthrough,
      seed: this.#seed,
      estimatorNames: this.#baseSpecs.map(s => s[0]),
      metaName: this.#metaSpec ? this.#metaSpec[0] : null,
    }
  }

  setParams(p) {
    this.#ensureAlive()
    if (p.cv !== undefined) this.#cv = p.cv
    if (p.passthrough !== undefined) this.#passthrough = p.passthrough
    if (p.seed !== undefined) this.#seed = p.seed
    return this
  }

  get capabilities() {
    return {
      classifier: this.#task === 'classification',
      regressor: this.#task === 'regression',
      predictProba: this.#task === 'classification',
      decisionFunction: false,
      sampleWeight: false,
      csr: false,
      earlyStopping: false,
    }
  }

  get isFitted() { return this.#fitted }
  get classes() { return this.#classes }

  // --- Private helpers ---

  #buildMetaFeatures(X) {
    const Xn = normalizeX(X)
    const n = Xn.rows
    const nBase = this.#baseModels.length
    const colsPerModel = this.#task === 'classification' ? this.#nClasses : 1
    const oofCols = nBase * colsPerModel

    const metaData = new Float64Array(n * this.#nMetaCols)
    for (let b = 0; b < nBase; b++) {
      if (this.#task === 'classification') {
        const proba = this.#baseModels[b].predictProba(Xn)
        for (let i = 0; i < n; i++) {
          for (let c = 0; c < this.#nClasses; c++) {
            metaData[i * this.#nMetaCols + b * colsPerModel + c] = proba[i * this.#nClasses + c]
          }
        }
      } else {
        const preds = this.#baseModels[b].predict(Xn)
        for (let i = 0; i < n; i++) {
          metaData[i * this.#nMetaCols + b] = preds[i]
        }
      }
    }

    if (this.#passthrough) {
      for (let i = 0; i < n; i++) {
        metaData.set(
          Xn.data.subarray(i * Xn.cols, (i + 1) * Xn.cols),
          i * this.#nMetaCols + oofCols
        )
      }
    }

    return { data: metaData, rows: n, cols: this.#nMetaCols }
  }

  static _register() {
    if (_registered) return
    _registered = true
    const loader = (manifest, toc, blobs) => {
      return StackingEnsemble._loadFromParts(manifest, toc, blobs)
    }
    register(TYPE_ID_CLS, loader)
    register(TYPE_ID_REG, loader)
  }

  static async _loadFromParts(manifest, toc, blobs) {
    const p = manifest.params
    const ens = new StackingEnsemble({
      task: p.task,
      cv: p.cv,
      passthrough: p.passthrough,
      seed: p.seed,
    })
    ens.#classes = p.classes ? new Int32Array(p.classes) : null
    ens.#nClasses = ens.#classes ? ens.#classes.length : 0
    ens.#nMetaCols = p.nMetaCols
    ens.#baseSpecs = p.estimatorNames.map(name => [name, null, null])
    ens.#metaSpec = [p.metaName, null, null]

    // Load base models
    ens.#baseModels = []
    for (const name of p.estimatorNames) {
      const entry = toc.find(t => t.id === name)
      if (!entry) throw new ValidationError(`No artifact for base estimator "${name}"`)
      const blob = blobs.subarray(entry.offset, entry.offset + entry.length)
      ens.#baseModels.push(await registryLoad(blob))
    }

    // Load meta-model
    const metaEntry = toc.find(t => t.id === p.metaName)
    if (!metaEntry) throw new ValidationError(`No artifact for meta estimator "${p.metaName}"`)
    const metaBlob = blobs.subarray(metaEntry.offset, metaEntry.offset + metaEntry.length)
    ens.#metaModel = await registryLoad(metaBlob)

    ens.#fitted = true
    return ens
  }
}

// --- Subset helpers ---

function _subsetX(X, indices) {
  const { data, cols } = X
  const rows = indices.length
  const out = new Float64Array(rows * cols)
  for (let i = 0; i < rows; i++) {
    const srcOff = indices[i] * cols
    out.set(data.subarray(srcOff, srcOff + cols), i * cols)
  }
  return { data: out, rows, cols }
}

function _subsetY(y, indices) {
  const out = new (y.constructor)(indices.length)
  for (let i = 0; i < indices.length; i++) {
    out[i] = y[indices[i]]
  }
  return out
}

File: py/wlearn/ensemble/_stacking.py
"""StackingEnsemble matching JS @wlearn/ensemble/stacking.js."""

import numpy as np

from ..errors import ValidationError, NotFittedError, DisposedError
from ..bundle import encode_bundle, decode_bundle
from ..registry import register, load as registry_load
from ..automl._cv import accuracy, r2_score, stratified_k_fold, k_fold

TYPE_ID_CLS = 'wlearn.ensemble.stacking.classifier@1'
TYPE_ID_REG = 'wlearn.ensemble.stacking.regressor@1'
_registered = False


class StackingEnsemble:
    def __init__(self, estimators=None, final_estimator=None, cv=5,
                 task='classification', passthrough=False, seed=42):
        """
        Args:
            estimators: list of (name, cls, params) tuples OR (name, fitted_model) tuples.
                If a tuple has 2 elements and the second is a fitted BaggedEstimator,
                its stored OOF predictions are used directly (no retraining).
            final_estimator: (name, cls, params) tuple for meta-model
            cv: number of folds
            task: 'classification' or 'regression'
            passthrough: include original features in meta features
            seed: random seed
        """
        self._base_specs = estimators or []
        self._meta_spec = final_estimator
        self._cv = cv
        self._task = task
        self._passthrough = passthrough
        self._seed = seed
        self._base_models = None
        self._meta_model = None
        self._classes = None
        self._n_classes = 0
        self._n_meta_cols = 0
        self._fitted = False
        self._disposed = False
        StackingEnsemble._register()

    @classmethod
    def create(cls, estimators=None, final_estimator=None, cv=5,
               task='classification', passthrough=False, seed=42):
        return cls(estimators=estimators, final_estimator=final_estimator,
                   cv=cv, task=task, passthrough=passthrough, seed=seed)

    def _ensure_alive(self):
        if self._disposed:
            raise DisposedError('StackingEnsemble has been disposed.')

    def _ensure_fitted(self):
        self._ensure_alive()
        if not self._fitted:
            raise NotFittedError('StackingEnsemble is not fitted. Call fit() first.')

    def fit(self, X, y):
        self._ensure_alive()
        if self._meta_spec is None:
            raise ValidationError('StackingEnsemble requires a finalEstimator')

        n = len(X)
        n_features = X.shape[1] if hasattr(X, 'shape') else len(X[0])

        if self._task == 'classification':
            labels = sorted(set(int(v) for v in y))
            self._classes = np.array(labels, dtype=np.int32)
            self._n_classes = len(self._classes)

        # Generate folds
        if self._task == 'classification':
            folds = stratified_k_fold(y, self._cv, do_shuffle=True, seed=self._seed)
        else:
            folds = k_fold(n, self._cv, do_shuffle=True, seed=self._seed)

        # Classify base specs: pre-fitted BaggedEstimator vs regular (name, cls, params)
        from ._bagging import BaggedEstimator
        bagged_bases = []  # (index, name, fitted_model) for pre-fitted BaggedEstimators
        spec_bases = []    # (index, name, cls, params) for regular specs
        for b, entry in enumerate(self._base_specs):
            if len(entry) == 2:
                name, model = entry
                if hasattr(model, 'oof_predictions') and model.is_fitted:
                    bagged_bases.append((b, name, model))
                else:
                    raise ValidationError(
                        f'Base estimator "{name}" is a 2-tuple but not a fitted '
                        f'BaggedEstimator with oof_predictions.'
                    )
            else:
                name, est_cls, params = entry
                spec_bases.append((b, name, est_cls, params))

        # Step 1: Generate OOF predictions
        n_base = len(self._base_specs)
        cols_per_model = self._n_classes if self._task == 'classification' else 1
        oof_cols = n_base * cols_per_model
        oof_data = np.zeros(n * oof_cols, dtype=np.float64)

        # Fill OOF from pre-fitted BaggedEstimators
        for b, name, model in bagged_bases:
            oof = model.oof_predictions
            if self._task == 'classification':
                nc = self._n_classes
                for i in range(n):
                    for c in range(nc):
                        oof_data[i * oof_cols + b * cols_per_model + c] = oof[i * nc + c]
            else:
                for i in range(n):
                    oof_data[i * oof_cols + b] = float(oof[i])

        # Generate OOF from regular specs via fold training
        for b, name, est_cls, params in spec_bases:
            for train, test in folds:
                X_train, y_train = X[train], y[train]
                X_test = X[test]

                model = est_cls.create(params or {})
                try:
                    model.fit(X_train, y_train)
                    if self._task == 'classification':
                        proba = model.predict_proba(X_test)
                        for i in range(len(test)):
                            row = test[i]
                            for c in range(self._n_classes):
                                oof_data[row * oof_cols + b * cols_per_model + c] = \
                                    proba[i * self._n_classes + c]
                    else:
                        preds = model.predict(X_test)
                        for i in range(len(test)):
                            oof_data[test[i] * oof_cols + b] = float(preds[i])
                finally:
                    model.dispose()

        # Step 2: Build meta-feature matrix
        if self._passthrough:
            self._n_meta_cols = oof_cols + n_features
            meta_data = np.zeros(n * self._n_meta_cols, dtype=np.float64)
            for i in range(n):
                meta_data[i * self._n_meta_cols:i * self._n_meta_cols + oof_cols] = \
                    oof_data[i * oof_cols:(i + 1) * oof_cols]
                meta_data[i * self._n_meta_cols + oof_cols:
                          i * self._n_meta_cols + oof_cols + n_features] = X[i]
            meta_X = meta_data.reshape(n, self._n_meta_cols)
        else:
            self._n_meta_cols = oof_cols
            meta_X = oof_data.reshape(n, oof_cols)

        # Step 3: Train base models on full data (or use pre-fitted BaggedEstimators)
        self._base_models = [None] * n_base
        for b, name, model in bagged_bases:
            self._base_models[b] = model
        for b, name, est_cls, params in spec_bases:
            model = est_cls.create(params or {})
            model.fit(X, y)
            self._base_models[b] = model

        # Step 4: Train meta-model on OOF features
        _, meta_cls, meta_params = self._meta_spec
        self._meta_model = meta_cls.create(meta_params or {})
        self._meta_model.fit(meta_X, y)

        self._fitted = True
        return self

    def predict(self, X):
        self._ensure_fitted()
        meta_X = self._build_meta_features(X)
        return self._meta_model.predict(meta_X)

    def predict_proba(self, X):
        self._ensure_fitted()
        if self._task != 'classification':
            raise ValidationError('predict_proba is only available for classification')
        if not hasattr(self._meta_model, 'predict_proba'):
            raise ValidationError('Meta-model does not support predict_proba')
        meta_X = self._build_meta_features(X)
        return self._meta_model.predict_proba(meta_X)

    def score(self, X, y):
        self._ensure_fitted()
        preds = self.predict(X)
        if self._task == 'classification':
            return accuracy(y, preds)
        return r2_score(y, preds)

    def save(self):
        self._ensure_fitted()
        type_id = TYPE_ID_CLS if self._task == 'classification' else TYPE_ID_REG
        manifest = {
            'typeId': type_id,
            'params': {
                'task': self._task,
                'cv': self._cv,
                'passthrough': self._passthrough,
                'seed': self._seed,
                'estimatorNames': [s[0] for s in self._base_specs],
                'metaName': self._meta_spec[0],
                'classes': list(self._classes) if self._classes is not None else None,
                'nMetaCols': self._n_meta_cols,
            },
        }
        artifacts = [
            {
                'id': self._base_specs[i][0],
                'data': self._base_models[i].save(),
                'mediaType': 'application/x-wlearn-bundle',
            }
            for i in range(len(self._base_models))
        ]
        artifacts.append({
            'id': self._meta_spec[0],
            'data': self._meta_model.save(),
            'mediaType': 'application/x-wlearn-bundle',
        })
        return encode_bundle(manifest, artifacts)

    @classmethod
    def load(cls, data):
        manifest, toc, blobs = decode_bundle(data)
        return cls._load_from_parts(manifest, toc, blobs)

    def dispose(self):
        if self._disposed:
            return
        self._disposed = True
        if self._base_models:
            for m in self._base_models:
                m.dispose()
        if self._meta_model:
            self._meta_model.dispose()

    def get_params(self):
        return {
            'task': self._task,
            'cv': self._cv,
            'passthrough': self._passthrough,
            'seed': self._seed,
            'estimatorNames': [s[0] for s in self._base_specs],
            'metaName': self._meta_spec[0] if self._meta_spec else None,
        }

    def set_params(self, p):
        self._ensure_alive()
        if 'cv' in p:
            self._cv = p['cv']
        if 'passthrough' in p:
            self._passthrough = p['passthrough']
        if 'seed' in p:
            self._seed = p['seed']
        return self

    @property
    def capabilities(self):
        return {
            'classifier': self._task == 'classification',
            'regressor': self._task == 'regression',
            'predictProba': self._task == 'classification',
            'decisionFunction': False,
            'sampleWeight': False,
            'csr': False,
            'earlyStopping': False,
        }

    @property
    def is_fitted(self):
        return self._fitted

    @property
    def classes(self):
        return self._classes

    def _build_meta_features(self, X):
        n = len(X)
        n_features = X.shape[1] if hasattr(X, 'shape') else len(X[0])
        n_base = len(self._base_models)
        cols_per_model = self._n_classes if self._task == 'classification' else 1
        oof_cols = n_base * cols_per_model

        meta_data = np.zeros(n * self._n_meta_cols, dtype=np.float64)
        for b in range(n_base):
            if self._task == 'classification':
                proba = self._base_models[b].predict_proba(X)
                for i in range(n):
                    for c in range(self._n_classes):
                        meta_data[i * self._n_meta_cols + b * cols_per_model + c] = \
                            proba[i * self._n_classes + c]
            else:
                preds = self._base_models[b].predict(X)
                for i in range(n):
                    meta_data[i * self._n_meta_cols + b] = float(preds[i])

        if self._passthrough:
            for i in range(n):
                for j in range(n_features):
                    meta_data[i * self._n_meta_cols + oof_cols + j] = X[i][j]

        return meta_data.reshape(n, self._n_meta_cols)

    @staticmethod
    def _register():
        global _registered
        if _registered:
            return
        _registered = True

        def loader(manifest, toc, blobs):
            return StackingEnsemble._load_from_parts(manifest, toc, blobs)

        register(TYPE_ID_CLS, loader)
        register(TYPE_ID_REG, loader)

    @staticmethod
    def _load_from_parts(manifest, toc, blobs):
        p = manifest['params']
        ens = StackingEnsemble(
            task=p['task'],
            cv=p.get('cv', 5),
            passthrough=p.get('passthrough', False),
            seed=p.get('seed', 42),
        )
        ens._classes = np.array(p['classes'], dtype=np.int32) if p.get('classes') else None
        ens._n_classes = len(ens._classes) if ens._classes is not None else 0
        ens._n_meta_cols = p.get('nMetaCols', 0)
        ens._base_specs = [(name, None, None) for name in p['estimatorNames']]
        ens._meta_spec = (p['metaName'], None, None)

        ens._base_models = []
        for name in p['estimatorNames']:
            entry = next((t for t in toc if t['id'] == name), None)
            if entry is None:
                raise ValidationError(f'No artifact for base estimator "{name}"')
            blob = bytes(blobs[entry['offset']:entry['offset'] + entry['length']])
            ens._base_models.append(registry_load(blob))

        meta_entry = next((t for t in toc if t['id'] == p['metaName']), None)
        if meta_entry is None:
            raise ValidationError(f'No artifact for meta estimator "{p["metaName"]}"')
        meta_blob = bytes(blobs[meta_entry['offset']:meta_entry['offset'] + meta_entry['length']])
        ens._meta_model = registry_load(meta_blob)

        ens._fitted = True
        return ens
